File: LeetCodeProblemsCMake/Problems/LeetCode3121CountTheNumberOfSpecialCharactersII.cpp
//
// Created by saeed on 5/27/26.
//

#include "LeetCode3121CountTheNumberOfSpecialCharactersII.h"

#include <unordered_map>
using namespace std;

int LeetCode3121CountTheNumberOfSpecialCharactersII::numberOfSpecialChars(string word) {
    unordered_map<char, int> charMaps;
    for (int i = 0, e = static_cast<int>(word.size()); i < e; i++) {
        if (word[i] > 'Z') {
            charMaps[word[i]] = i;
            continue;
        }
        if (charMaps.contains(word[i]))
            continue;
        charMaps[word[i]] = i;
    }
    int result{};
    for (int i = 'a'; i <= 'z'; i++) {
        if (!charMaps.contains(i))
            continue;
        if (!charMaps.contains(i - 'a' + 'A'))
            continue;
        if (charMaps[i] > charMaps[i - 'a' + 'A'])
            continue;
        ++result;
    }
    return result;
}

File: LeetCodeProblemsGo/LeetCode3121CountTheNumberOfSpecialCharactersII/LeetCode3121CountTheNumberOfSpecialCharactersII.go
package LeetCode3121CountTheNumberOfSpecialCharactersII

func numberOfSpecialChars(word string) int {
	upperCase := make(map[rune]int)
	lowerCase := make(map[rune]int)
	for i, e := 0, len(word); i < e; i++ {
		if word[i] > 'Z' {
			lowerCase[rune(word[i])] = i
			continue
		}
		if _, ok := upperCase[rune(word[i])]; ok {
			continue
		}
		upperCase[rune(word[i])] = i
	}
	result := 0
	for i := 'a'; i <= 'z'; i++ {
		if _, ok := lowerCase[i]; !ok {
			continue
		}
		if _, ok := upperCase[i-'a'+'A']; !ok {
			continue
		}
		if lowerCase[i] > upperCase[i-'a'+'A'] {
			continue
		}
		result++
	}
	return result
}
